File: openc3/lib/openc3/interfaces/protocols/cmd_response_protocol.rb
# encoding: ascii-8bit

# Copyright 2024 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/config/config_parser'
require 'openc3/interfaces/protocols/protocol'
require 'thread' # For Queue
require 'timeout' # For Timeout::Error

module OpenC3
  # Protocol that waits for a response for any commands with a defined response packet.
  # The response packet is identified but not defined by the protocol.
  class CmdResponseProtocol < Protocol
    # @param response_timeout [Float] Number of seconds to wait before timing out
    #   when waiting for a response
    # @param response_polling_period [Float] Number of seconds to wait between polling
    #   for a response
    # @param raise_exceptions [String] Whether to raise exceptions when errors
    #   occur in the protocol like unexpected responses or response timeouts.
    # @param allow_empty_data [true/false/nil] See Protocol#initialize
    def initialize(
      response_timeout = 5.0,
      response_polling_period = 0.02,
      raise_exceptions = false,
      allow_empty_data = nil
    )
      super(allow_empty_data)
      @response_timeout = ConfigParser.handle_nil(response_timeout)
      @response_timeout = @response_timeout.to_f if @response_timeout
      @response_polling_period = response_polling_period.to_f
      @raise_exceptions = ConfigParser.handle_true_false(raise_exceptions)
      @write_block_queue = Queue.new
      @response_packet = nil
    end

    def connect_reset
      super()
      begin
        @write_block_queue.pop(true) while @write_block_queue.length > 0
      rescue
      end
    end

    def disconnect_reset
      super()
      @write_block_queue << nil # Unblock the write block queue
    end

    def read_packet(packet)
      if @response_packet
        # Grab the response packet specified in the command
        result_packet = System.telemetry.packet(@response_packet[0], @response_packet[1]).clone
        result_packet.buffer = packet.buffer
        result_packet.received_time = nil
        result_packet.stored = packet.stored
        result_packet.extra = packet.extra

        # Release the write
        @write_block_queue << nil

        # This returns the fully identified and defined packet
        # Received time is handled by the interface microservice
        return result_packet
      else
        return packet
      end
    end

    def write_packet(packet)
      # Setup the response packet (if there is one)
      # This primes waiting for the response in post_write_interface
      @response_packet = packet.response

      return packet
    end

    def post_write_interface(packet, data, extra = nil)
      if @response_packet
        if @response_timeout
          response_timeout_time = Time.now + @response_timeout
        else
          response_timeout_time = nil
        end

        # Block the write until the response is received
        begin
          @write_block_queue.pop(true)
        rescue
          sleep(@response_polling_period)
          retry if !response_timeout_time
          retry if response_timeout_time and Time.now < response_timeout_time
          handle_error("#{@interface ? @interface.name : ""}: Timeout waiting for response")
        end

        @response_packet = nil
      end
      return super(packet, data, extra)
    end

    def handle_error(msg)
      Logger.error(msg)
      raise msg if @raise_exceptions
    end
  end
end

File: openc3/python/openc3/interfaces/protocols/cmd_response_protocol.py
# Copyright 2024 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

from openc3.system.system import System
from openc3.config.config_parser import ConfigParser
from openc3.utilities.logger import Logger
from openc3.interfaces.protocols.protocol import Protocol
from queue import SimpleQueue, Empty
import time


# Protocol that waits for a response for any commands with a defined response packet.
# The response packet is identified but not defined by the protocol.
class CmdResponseProtocol(Protocol):
    # @param response_timeout [Float] Number of seconds to wait before timing out
    #   when waiting for a response
    # @param response_polling_period [Float] Number of seconds to wait between polling
    #   for a response
    # @param raise_exceptions [String] Whether to raise exceptions when errors
    #   occur in the protocol like unexpected responses or response timeouts.
    # @param allow_empty_data [true/false/nil] See Protocol#initialize
    def __init__(
        self, response_timeout=5.0, response_polling_period=0.02, raise_exceptions=False, allow_empty_data=None
    ):
        super().__init__(allow_empty_data)
        self.response_timeout = ConfigParser.handle_none(response_timeout)
        if self.response_timeout:
            self.response_timeout = float(response_timeout)
        self.response_polling_period = float(response_polling_period)
        self.raise_exceptions = ConfigParser.handle_true_false(raise_exceptions)
        self.write_block_queue = SimpleQueue()
        self.response_packet = None

    def connect_reset(self):
        super().connect_reset()
        try:
            while self.write_block_queue.qsize != 0:
                self.write_block_queue.get_nowait()
        except Empty:
            pass

    def disconnect_reset(self):
        super().disconnect_reset()
        self.write_block_queue.put(None)  # Unblock the write block queue

    def read_packet(self, packet):
        if self.response_packet is not None:
            # Grab the response packet specified in the command
            result_packet = System.telemetry.packet(self.response_packet[0], self.response_packet[1]).clone()
            result_packet.buffer = packet.buffer
            result_packet.received_time = None
            result_packet.stored = packet.stored
            result_packet.extra = packet.extra

            # Release the write
            self.write_block_queue.put(None)

            # This returns the fully identified and defined packet
            # Received time is handled by the interface microservice
            return result_packet
        else:
            return packet

    def write_packet(self, packet):
        # Setup the response packet (if there is one)
        # This primes waiting for the response in post_write_interface
        self.response_packet = packet.response

        return packet

    def post_write_interface(self, packet, data, extra=None):
        if self.response_packet is not None:
            if self.response_timeout:
                response_timeout_time = time.time() + self.response_timeout
            else:
                response_timeout_time = None

            # Block the write until the response is received
            while True:
                try:
                    self.write_block_queue.get_nowait()
                    break
                except Empty:
                    time.sleep(self.response_polling_period)
                    if response_timeout_time is None:
                        continue
                    if response_timeout_time and time.time() < response_timeout_time:
                        continue
                    interface_name = ""
                    if self.interface is not None:
                        interface_name = self.interface.name
                    self.handle_error(f"{interface_name}: Timeout waiting for response")

            self.response_packet = None
        return super().post_write_interface(packet, data, extra)

    def handle_error(self, msg):
        Logger.error(msg)
        if self.raise_exceptions:
            raise RuntimeError(msg)
